### DISCUSSION TITLE
A painful issue with corrupted ActiveStorageAttachment's

### DISCUSSION THREAD

**Discussion by ok32** [solidusio/solidus#3986] [Troubleshooting]
Hey guys!
I have an issue using `ActiveStorageAttachment` for product images.
I'm using an S3-compatible service for storing attachments, and it's working ok. But if the service returns an error (it is down at the moment the user attempts to upload an image or there's a permission error, etc), a product image record and ActiveStorage blob record are created anyway despite the image processing error. So it produces a corrupted state of the system. And then any code which calls `.url(style)` on this image gets an `ActiveStorage::FileNotFoundError` exception, because it tries to run image processing to create the requested image variant, but it can't find the original image because its upload never finished (it happens on any page which needs to display the image).
Ideally, I would like it to be "all or nothing", so the database records get committed only after the storage service responded with a success.
Is it a problem of how ActiveStorage works itself or is it a Solidus wrapper's one? How do I prevent the situation when one failed image upload breaks down the whole store (you can't even delete the corrupted image, because the admin product page is also broken)?

**Comment by elia** [comment]
@ok32 can you post some more info on where/how it fails? That would be very helpful in trying to understand where is the problem 🙏 

_This might be related: https://github.com/rails/rails/issues/31985_

**Reply by ok32** [reply]
It fails with an ActiveStorage exception on any page which initiates ActiveStorage attachment processing (generation of image variants in this case).

From our discussion with @kennyadsl on Slack:

> kennyadsl:nebulab:
> I think we have to kinda rescue that error in Solidus and gracefully emit some error. I’ve seen something similar trying to upload a non supported format. I had to remove that corrupted image manually from the database. I still don’t have the answer, have to investigate better.

> Evgeny:
> As a workaround I copied Spree::Image::ActiveStorageAttachment to my app and wrapped the url() method (and made Solidus use it):

```
module Spree::Image::MyActiveStorageAttachment
  extend ActiveSupport::Concern
  include Spree::ActiveStorageAdapter
  delegate :width, :height, to: :attachment, prefix: true
  included do
    has_attachment :attachment,
                   styles: {
                   mini: '48x48>',
                   small: '400x400>',
                   product: '680x680>',
                   large: '1200x1200>'
                 },
                 default_style: :product
    validates :attachment, presence: true
    validate :attachment_is_an_image

    def url(style = nil)
      super(style)
    rescue ActiveStorage::FileNotFoundError
      '/images/corrupted-image.png'
    end

  end
end
```

**Comment by kennyadsl** [comment]
I think I was able to reproduce this by uploading a `.heic` file.

**Comment by kennyadsl** [comment]
I've created https://github.com/solidusio/solidus/issues/4017. This is a bug and we should fix it before releasing Solidus 3.0, which has Active Storage set as the default adapter.

**Comment by kennyadsl** [comment]
Hey, this is solved thanks to @cpfergus1 with https://github.com/solidusio/solidus/pull/4021 and https://github.com/solidusio/solidus/pull/4026. 🎉
